### DISCUSSION TITLE
How Refnotes work for bibliography

### DISCUSSION THREAD

**Discussion by kookma** [kookma/Refnotes#8] [Show and tell]
**Refnotes 1.0.3 of 2020**

Asked by Douglas at [Google Forum](https://groups.google.com/d/msgid/tiddlywiki/09b464b5-6608-4368-9105-36615966f0d6n%40googlegroups.com.) How Refnotes work for bibliography?

Hi Douglas,

 I will hopefully work on Refnotes documentation to help newbies find their way easily. But just to answer your questions above

1. Refnotes has three small tools
1.1. abbreviations (to insert abbreviations and create glossary table)
1.2. footnotes (to insert numbered footnote in the tiddler body and create at the bottom of the tiddler full length footnotes)
1.3. bibliography (to insert references, and create bibliography table)

To use it:

1. drag and drop the Refnotes to your wiki
2. install Bibtex importer (not required but automate things)

i. Normally people use some libraries to get the bibliography data. Almost any library support bibtex format, so you need to export bibliography data as .bib files
ii. import the bibliography data (.bib files to your wiki. Bibtex Importer gets the data and generates bibtex entry tiddlers with all required fields (automatically)
iii. you can now cite using simple Refnotes macro

That's all. 

Refnotes plugin does not manually create bibtex entries nor encourage to do that! Bibliography data can be retrieved from a library like Google Scholar, sciencedirect.com, library of congress, ... 

But as an enhancement, I can add a macro to get data from the user and all the user to create a bibtex entry manually 

Hope this helps you and answers some of your questions.

**Comment by kookma** [comment]
Replied by [Douglas Glenn](https://groups.google.com/g/tiddlywiki/c/b09whBNaO3w/m/5te627pqBAAJ)

re-populating the user fields, ect. was a thought of the moment, but I agree with you on pre populating fields. I suggest as an alternative instead of spending the time and energy to build an interface in Refnotes. Just add an enhancement to import .bib files, and provide a short list of links to editors so you don't have to reinvent the wheel. I also included on tutorial site as well so rank beginners such as myself can use to self-educate from an easy to read environment. I wouldn't recommend people to have to dig into the actual documentation. It's overkill and more likely to turn people off when they see it.  I hope this helps! and thank you for taking the time to listen to me.

Tutorial:
    https://www.latex-tutorial.com/tutorials/bibtex/

Free Online BibTEX editor:   https://truben.no/latex/bibtex/

Free Cross platform editors:  
    https://www.jabref.org  (Apple OS X, Windows, Linux)
    https://miktex.org/

Native Linux
Linux KDE gui:  bibTEX (through the package manager for the distro)

Linux Gnome gui:  
    Setzer (https://github.com/GNOME/gnome-latex), 
    Marker (https://github.com/fabiocolacio/Marker), 

Aggregate listings of multiple editors in  a single place   
    10 Best Latex editors for 2021 :   https://beebom.com/best-latex-editors/
    https://botw.org/top/Computers/Software/Typesetting/TeX/BibTeX/

**Comment by rahulkashyap411** [comment]
Dear Mohammad,

I am working in academia (Physics). 
I was really impressed by this plugin. It was until I saw this that I found the bibtex plugin useful.  I really look forward to future improvements. 
I do have a suggestion/request. I would also love to link the local pdf of references so that when I hover over the reference, I can click the link and open the pdf right away. 
So far I have been using a python script to create the tiddlers from my literature folder; here is my script -- https://github.com/rahulkashyap411/tw_scripts/blob/master/update_journalArticle_tid.py 

Do you think it's be possible? Is this a change that could be done at Refnote plugin or, Bibtex plugin or, perhaps both? 

Looking forward for your reply. 

-Rahul

**Reply by kookma** [reply]
@rahulkashyap411 - thank you for your suggestion. I add this as an enhancement and I will see how can I implement. I will then ping you here! Just I will come back to this hopefully in early Feb.

**Reply by rahulkashyap411** [reply]
Thanks for responding. I wanted to update my thoughts on this -- 

* I have bibtex file where each entry also contains a field called, "bibtex-file"; using this I can get the full path to the pdf file. I checked that the Bibtex importer plugin also imports this in a field with same name in the tiddler specific to each entry.  

*  Another thought I had was to display a description different than the cite-key to the references when you hover mouse over them; This is more than beautification and brings clarity to the reference. 

I believe that Latex does not represent the non-linear structure of knowledge as tiddlywiki does. However,  I'm basically trying to get a document with best features of formatting as close to latex as possible. 

Aside topic: 
I am also thinking about auto-numbering of sections (sub-sections), equations, figures, tables and lists. All of these has to be consistent especially when you transclude otherwise we're losing the most powerful feature of tiddlywiki.

**Reply by kookma** [reply]
@rahulkashyap411 - thank you for your feedback. I think there is a plugin do the work, I will have a look in the forum to see what is the name. I will start working on Refnotes in Feb and your comments are highly welcome

**Reply by rahulkashyap411** [reply]
Dear Mohammad,

Many thanks for your updates on refnotes. 

I tried to implement what I was looking for but, couldn't do. Since you're familiar with this, I think it'd be easier for you to suggest me. I believe it's the macro displayref-onhover where I need to put the link using [ext[pdf|..... put here the content of bibtex-file.... ]] [[url|...put here the content of bibtex-url]] 

I'd also like to include them in the showrefs macro.  

This is how I tried to modifyit. It shows the text "pdf, url" but, clicking it doesn't get me to pdf or url and prompt to create new tiddler. 

\define displayref-onhover(refTid)
<$set name="ref-tid" value=<<__refTid__>> >
<$link overrideClass="link-refcls" to=<<ref-tid>> >
<$list filter="[<ref-tid>get[bibtex-entry-type]lowercase[]!match[website]]" variable=null
emptyMessage="""<span class="ref-author"><$view tiddler=<<ref-tid>> field="bibtex-url"/>.</span>""">
<span class="ref-author"><$view tiddler=<<ref-tid>> field="bibtex-author"/>.</span>
</$list>
</$link>
<$view tiddler=<<ref-tid>> field="bibtex-title"/>. (<$view tiddler=<<ref-tid>> field="bibtex-year"/>, **_[ext[pdf|<$view field="bibtex-file" tiddler=<<curtid>>/>]] [[url|<$view tiddler=<<ref-tid>> field="bibtex-url"/>]]_** )
</$set>
\end


Regards,
Rahul

**Reply by kookma** [reply]
Hi Rahul,

First part
Where are the pdfs? Have you checked the links to see if they work correctly?
I may ask to have a test tiddler and using `[ext[pdf-url]]` check to see the links are not broken!

The next part
the pdf links need to be added to the bibtex entry e.g add a field like bibtex-pdf-url

The last part:
As these are external links you have to use `a` tag like below:  (see https://www.w3schools.com/tags/tag_a.asp)

```
<a href={{{ [<refTid>get[bibtex-pdf-url]] }}} target=_blank>pdf </>
```

**Reply by rahulkashyap411** [reply]
thanks for quick reply. 

pdfs are in a local folder and when I give their paths in the tiddler main text -- [ext[pdf|bibtex-file]], it works. 

I do have these two fileds in my bibtex entries from which I can get relevant details. 

**bibtex-file** : contains relative path to pdf; want to include it like [ext[pdf|bibtex-file]] 
**bibtex-url**: contains web address; [[url|bibtex-url]] 

I checked your method; doesn't seem to work.

**Reply by kookma** [reply]
In a test tiddler put the below code and see how it works

```
<$tiddler tiddler="name-of-tiddler-contains-bibtex-entry ">
<a class="tc-tiddlylink-external" href={{{ [<currentTiddler>get[bibtex-pdf]] }}} 
   target="_blank" rel="noopener noreferrer"> pdf  </a>
</$tiddler>
```

Assuming 
- name-of-tiddler-contains-bibtex-entry contains a field with url to pdf
- the pdf url is stored in `bibtex-pdf`

The above code works for me! 

Note the bibtex-pdf is the name of pdf if it is in the same folder of wiki! or a correct url if it is on the server

**Reply by rahulkashyap411** [reply]
Many thanks for this. It's working now. 
I had to include the whole thing in the displayref-onhover. 

Two quick questions -- 

1. Can I process the string in the "bibtex-file" field to add a modification of it such as trim it and append with another string in front of it?  Could you please point me towards a reference for this? 
2. I also want to do the same for the page-end reference list. Which function does that? Is it **$:/plugins/kookma/refnotes/macros/apa/showrefs**

Thanks again,
Rahul

**Reply by kookma** [reply]
> * Can I process the string in the "bibtex-file" field to add a modification of it such as trim it and append with another string in front of it?  Could you please point me towards a reference for this?

Yes! one solution is to use a $vars! for example you want to add a prefix like `/assets/pdfs/` to the `bibtex-file` field

```
<$vars pdfu = {{{ [<refTid>get[bibtex-pdf]addprefix[/asset/pdfs/]]] }}}>
..
```

> * I also want to do the same for the page-end reference list. Which function does that? Is it **$:/plugins/kookma/refnotes/macros/apa/showrefs**

The showrefs uses templates to display the bibliographies. So open the proper template for example template for book ($:/plugins/kookma/refnotes/templates/reflist/book) or article ($:/plugins/kookma/refnotes/templates/reflist/article) and add change it as you like.
